napari/plugins/_npe2.py

```diff
@@ -282,11 +296,105 @@ def _on_plugin_enablement_change(enabled: Set[str], disabled: Set[str]):
+        # currently registered/available.  So we check to make sure this is
+        # actually a registered plugin.
+        if plugin_name in pm.instance():
+            _register_manifest_actions(pm.get_manifest(plugin_name))
+
+    # TODO: after app-model, these QMenus will be evented and self-updating
+    # and we can remove this... but `_register_manifest_actions` will need to
+    # add the actions file and plugins menus (since we don't require plugins to
+    # list them explicitly)
     for v in Viewer._instances:
         v.window.plugins_menu._build()
         v.window.file_menu._rebuild_samples_menu()
 
 
-def index_npe1_adapters():
-    """Tell npe2 to import and index any discovered npe1 plugins."""
-    pm.index_npe1_adapters()
+def on_plugins_registered(manifests: Set[PluginManifest]):
+    """Callback when any npe2 plugins are registered.
+
+    'Registered' means that a manifest has been provided or discovered.
+    """
+    for mf in manifests:
+        if not pm.is_disabled(mf.name):
+            _register_manifest_actions(mf)
+
+
+def _register_manifest_actions(manifest: PluginManifest) -> None:
+    """Gather and register actions from a manifest.
+
+    This is called when a plugin is registered or enabled and it adds the
+    plugin's menus and submenus to the app model registry.
+    """
+    from .._app_model import get_app
+
+    app = get_app()
+    actions, submenus = _npe2_manifest_to_actions(manifest)
+    context = pm.get_context(cast('PluginName', manifest.name))
+    if actions:
+        context.register_disposable(app.register_actions(actions))
+    if submenus:
+        context.register_disposable(app.menus.append_menu_items(submenus))
+
+
+def _npe2_manifest_to_actions(
+    mf: PluginManifest,
+) -> Tuple[List[Action], List[Tuple[str, SubmenuItem]]]:
+    """Gather actions and submenus from a npe2 manifest, export app_model types."""
+    from app_model.types import Action, MenuRule
+
+    from .._app_model.constants._menus import is_menu_contributable
+
+    cmds: DefaultDict[str, List[MenuRule]] = DefaultDict(list)
+    submenus: List[Tuple[str, SubmenuItem]] = []
+    for menu_id, items in mf.contributions.menus.items():
+        if is_menu_contributable(menu_id):
+            for item in items:
+                if isinstance(item, contributions.MenuCommand):
+                    rule = MenuRule(id=menu_id, **_when_group_order(item))
+                    cmds[item.command].append(rule)
+                else:
+                    subitem = _npe2_submenu_to_app_model(item)
+                    submenus.append((menu_id, subitem))
+
+    actions: List[Action] = [
+        Action(
+            id=cmd.id,
+            title=cmd.title,
+            category=cmd.category,
+            tooltip=cmd.short_title or cmd.title,
+            icon=cmd.icon,
+            enablement=cmd.enablement,
+            callback=cmd.python_name or '',
+            menus=cmds.get(cmd.id),
+            keybindings=[],
+        )
```

Comment:
npe2 does not allow for togglable actions?

Reply:
nope, it's not in the current model